File: backend/functions/src/middleware/authMiddleware.ts
import { Request, Response, NextFunction } from "express";
import * as admin from "firebase-admin";

// Extend Express Request to carry userId downstream
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Express {
    interface Request {
      userId?: string;
    }
  }
}

/**
 * Verifies the Firebase ID token sent in the Authorization header.
 * Attaches req.userId on success.
 *
 * Clients must send: Authorization: Bearer <firebase-id-token>
 */
export const verifyAuth = async (
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> => {
  const authHeader = req.headers.authorization;

  if (!authHeader || !authHeader.startsWith("Bearer ")) {
    res.status(401).json({ error: "Unauthorized: missing Bearer token." });
    return;
  }

  const idToken = authHeader.split("Bearer ")[1];

  try {
    const decodedToken = await admin.auth().verifyIdToken(idToken);
    req.userId = decodedToken.uid;
    next();
  } catch (error) {
    console.error("Auth token verification failed:", error);
    res.status(401).json({ error: "Unauthorized: invalid or expired token." });
  }
};

File: backend/lib/core/auth_middleware.dart
import 'package:shelf/shelf.dart';
import 'firebase_setup.dart';

Middleware authMiddleware() {
  return (Handler innerHandler) {
    return (Request request) async {
      // OPTIONS requests should be passed through for CORS
      if (request.method == 'OPTIONS') {
        return innerHandler(request);
      }

      final authHeader = request.headers['authorization'] ?? request.headers['Authorization'];
      
      if (authHeader == null || !authHeader.startsWith('Bearer ')) {
        return Response(401, body: 'Missing or invalid Authorization header');
      }

      final token = authHeader.substring(7);

      try {
        final decodedToken = await AppSetup.auth.verifyIdToken(token);
        // We can add the user to the request context
        final updatedRequest = request.change(context: {'userId': decodedToken.uid});
        return innerHandler(updatedRequest);
      } catch (e) {
        return Response(401, body: 'Unauthorized: Invalid token');
      }
    };
  };
}
